tests/0098-consumer-txn.cpp

```diff
@@ -0,0 +1,553 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <iostream>
+#include <cstring>
+#include <cstdlib>
+#include "testcpp.h"
+#include <assert.h>
+#include <sstream>
+
+
+/**
+ * @name Consumer Transactions.
+ *
+ * - Uses the TransactionProducerCli Java application to produce messages
+ *   that are part of abort and commit transactions in various combinations
+ *   and tests that librdkafka consumes them as expected. Refer to 
+ *   TransactionProducerCli.java for scenarios covered.
+ */
+
+
+static void test_assert(bool cond, std::string msg) {
+  if (!cond)
+    Test::Say(msg);
+  assert(cond);
+}
+
+
+static void execute_java_produce_cli(std::string &bootstrapServers, std::string &topic, std::string cmd) {
+  std::stringstream ss;
+  ss << "./java/run-class.sh TransactionProducerCli " + bootstrapServers + " " + topic + " " + cmd;
+  int status = system(ss.str().c_str());
+  test_assert(!status, "./java/run-class.sh TransactionProducerCli failed with error code: " + status);
+}
+
+static std::vector<RdKafka::Message *> consume_messages(RdKafka::KafkaConsumer *c, std::string topic, int partition) {
+  RdKafka::ErrorCode err;
+
+  /* Assign partitions */
+  std::vector<RdKafka::TopicPartition*> parts;
+  parts.push_back(RdKafka::TopicPartition::create(topic, partition));
+  if ((err = c->assign(parts)))
+    Test::Fail("assign failed: " + RdKafka::err2str(err));
+  RdKafka::TopicPartition::destroy(parts);
+
+  Test::Say("Consuming from topic " + topic + "\n");  
+  std::vector<RdKafka::Message *> result = std::vector<RdKafka::Message *>();
+
+  while (true) {
+    RdKafka::Message *msg = c->consume(tmout_multip(1000));
+    switch (msg->err())
+    {
+      case RdKafka::ERR__TIMED_OUT:
+        continue;
+      case RdKafka::ERR__PARTITION_EOF:
```

Comment:
I'd like to see a test for PARTITION_EOF in two modes where the LSO is < HWMARK, one for committed and one for uncommitted, to verify that it behaves according to the isolation.level.

Reply:
yep, good plan.

Reply:
note that every test uses the `consume_messages` function which relies on partition eof events to know when to finish and every test is run for each isolation mode. I believe the scenario you suggest should already be covered, with the caveat that i observed producers don't necessarily send aborted transactions, though i recall making a scenario where they do. will check.

Reply:
good call, this is broken.

Reply:
also added a check for ls_offset and hi_offset in the stats. the code is a bit less than ideal, but you can feel free to not care about that.